### ISSUE TITLE
Escaped Single-Quote will fail to Lex

### ISSUE THREAD

**Issue by raphael-riel** [sqlfluff/sqlfluff#1081]
-- If this is a parsing or linting issue, please include a minimal SQL example which reproduces the issue, along with the `sqlfluff parse` output, `sqlfluff lint` output and `sqlfluff fix` output when relevant. Finally, please add any labels that are related to the issue before submitting so that it can be addressed easily! --

## Expected Behaviour

I would expect this snippet to properly lex.
```sql
select
    'I\'ll fail to Lex' as failing_string
from latest

```

## Observed Behaviour

On 0.5.3 I get:

```
=== [dbt templater] Compiling dbt project...
== [models/staging/backend/stg_backend_licenses.sql] FAIL
L:   2 | P:  23 |  LXR | Unable to lex characters: '"' as faili"...'
```

On 0.5.6:

```
=== [dbt templater] Compiling dbt project...
WARNING    Zero length Lex item returned from '<unlexable>'. Report this as a bug. 
== [models/staging/backend/stg_backend_licenses.sql] FAIL
L:   0 | P:   0 |  LXR | Fatal. Unable to lex characters: 0
```

## Version
`0.5.6` (Also tested with `0.5.3`)

## Configuration
```
[sqlfluff]
templater = dbt
dialect = snowflake

exclude_rules = L009,L031,L034,L044

[sqlfluff:rules:L016]
max_line_length=160

[sqlfluff:rules:L014]
extended_capitalisation_policy=lower

```

**Comment by raphael-riel** [comment]
I'm able to get SQLFluff to LEX the file by escaping the string using Snowflake's `$$` syntax.
https://docs.snowflake.com/en/sql-reference/data-types-text.html#dollar-quoted-string-constants

```sql
select
    $$I\'ll fail to Lex$$ as failing_string
from latest
```

will yield:

```
=== [dbt templater] Compiling dbt project...
== [models/staging/backend/stg_backend_licenses.sql] FAIL
L:   1 | P:   1 | L036 | Select targets should be on a new line unless there is
                       | only one select target.
L:   2 | P:   5 |  PRS | Found unparsable section: "$$I\\'ll fail to Lex$$ as
                       | failing_string"

Process finished with exit code 65
```

Which is slightly better. 
Should I open an issue about supporting this Snowflake's syntax?


Ignoring the Unparsable section:

```sql
select
    $$I\'ll fail to Lex$$ as failing_string  -- noqa: PRS
from latest
```
Yields:
```
=== [dbt templater] Compiling dbt project...
== [models/staging/backend/stg_backend_licenses.sql] FAIL
L:   1 | P:   1 | L036 | Select targets should be on a new line unless there is
                       | only one select target.
L:   2 | P:  46 | L003 | Comment not aligned with following line.
```

Which allows me to move forward 🎉 

Thanks for doing such greatly executed piece of software! Would not be able to continue without all of the nice configurations and features!

**Comment by tunetheweb** [comment]
[Snowflake documentation](https://docs.snowflake.com/en/sql-reference/data-types-text.html#single-quoted-string-constants) says:
> To include a single quote character within a string constant, type two adjacent single quotes (e.g. '').

And give the example of:

```sql
select 'Today''s sales projections', '-''''-';
```

We do explicitly support this:

https://github.com/sqlfluff/sqlfluff/blob/e21d200c35b75d771f2d7635460f9c0e7d9d6f06/src/sqlfluff/dialects/dialect_snowflake.py#L36-L42

So I think this can be closed. Feel free to reopen if that's incorrect.
